File: results/0.75/JPEG/jpeg-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define SOI 0xD8FF
#define APP0 0xE0FF
#define APP1 0xE1FF
#define DQT 0xDBFF
#define SOF0 0xC0FF
#define DHT 0xC4FF
#define SOS 0xDAFF
#define EOI 0xD9FF
#define DRI 0xDDFF
#define COM 0xFEFF
#define APP2 0xE2FF
#define APP3 0xE3FF
#define APP4 0xE4FF
#define APP5 0xE5FF
#define APP6 0xE6FF
#define APP7 0xE7FF
#define APP8 0xE8FF
#define APP9 0xE9FF
#define APP10 0xEAFF
#define APP11 0xEBFF
#define APP12 0xECFF
#define APP13 0xEDFF
#define APP14 0xEEFF
#define APP15 0xEFFF

typedef struct {
    uint16_t length;
    uint8_t identifier[5];
    uint16_t version;
    uint8_t units;
    uint16_t xdensity;
    uint16_t ydensity;
    uint8_t thumb_width;
    uint8_t thumb_height;
} app0_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app1_t;

typedef struct {
    uint16_t length;
    uint8_t table_number;
    uint8_t precision;
    uint8_t table_data[0];
} dqt_t;

typedef struct {
    uint16_t length;
    uint8_t precision;
    uint16_t image_height;
    uint16_t image_width;
    uint8_t number_of_components;
    uint8_t component_identifier;
    uint8_t horizontal_sampling_factor;
    uint8_t vertical_sampling_factor;
    uint8_t quantization_table_number;
} sof0_t;

typedef struct {
    uint16_t length;
    uint8_t table_class;
    uint8_t table_number;
    uint8_t table_data[0];
} dht_t;

typedef struct {
    uint16_t length;
    uint8_t number_of_components;
    uint8_t component_identifier;
    uint8_t dc_entropy_coding;
    uint8_t ac_entropy_coding;
} sos_t;

typedef struct {
    uint16_t length;
    uint16_t restart_interval;
} dri_t;

typedef struct {
    uint16_t length;
    uint8_t comment[0];
} com_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app2_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app3_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app4_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app5_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app6_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app7_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app8_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app9_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app10_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app11_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app12_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app13_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app14_t;

typedef struct {
    uint16_t length;
    uint8_t identifier[0];
} app15_t;

int main(int argc, char **argv) {
    if (argc < 2) {
        printf("Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        printf("Error opening file %s\n", argv[1]);
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *data = malloc(file_size);
    if (!data) {
        printf("Error allocating memory\n");
        fclose(file);
        return 1;
    }

    size_t bytes_read = fread(data, 1, file_size, file);
    if (bytes_read != file_size) {
        printf("Error reading file %s\n", argv[1]);
        free(data);
        fclose(file);
        return 1;
    }

    fclose(file);

    HParser *soi_parser = h_word_u16be(SOI);
    HParser *app0_parser = h_struct(
        app0_t,
        h_word_u16be(APP0),
        h_word_u16be(),
        h_bytes(5),
        h_word_u16be(),
        h_byte(),
        h_word_u16be(),
        h_word_u16be(),
        h_byte(),
        h_byte()
    );

    HParser *app1_parser = h_struct(
        app1_t,
        h_word_u16be(APP1),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *dqt_parser = h_struct(
        dqt_t,
        h_word_u16be(DQT),
        h_word_u16be(),
        h_byte(),
        h_byte(),
        h_bytes_until(h_any())
    );

    HParser *sof0_parser = h_struct(
        sof0_t,
        h_word_u16be(SOF0),
        h_word_u16be(),
        h_byte(),
        h_word_u16be(),
        h_word_u16be(),
        h_byte(),
        h_byte(),
        h_byte(),
        h_byte(),
        h_byte()
    );

    HParser *dht_parser = h_struct(
        dht_t,
        h_word_u16be(DHT),
        h_word_u16be(),
        h_byte(),
        h_byte(),
        h_bytes_until(h_any())
    );

    HParser *sos_parser = h_struct(
        sos_t,
        h_word_u16be(SOS),
        h_word_u16be(),
        h_byte(),
        h_byte(),
        h_byte(),
        h_byte()
    );

    HParser *eoi_parser = h_word_u16be(EOI);

    HParser *dri_parser = h_struct(
        dri_t,
        h_word_u16be(DRI),
        h_word_u16be(),
        h_word_u16be()
    );

    HParser *com_parser = h_struct(
        com_t,
        h_word_u16be(COM),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app2_parser = h_struct(
        app2_t,
        h_word_u16be(APP2),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app3_parser = h_struct(
        app3_t,
        h_word_u16be(APP3),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app4_parser = h_struct(
        app4_t,
        h_word_u16be(APP4),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app5_parser = h_struct(
        app5_t,
        h_word_u16be(APP5),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app6_parser = h_struct(
        app6_t,
        h_word_u16be(APP6),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app7_parser = h_struct(
        app7_t,
        h_word_u16be(APP7),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app8_parser = h_struct(
        app8_t,
        h_word_u16be(APP8),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app9_parser = h_struct(
        app9_t,
        h_word_u16be(APP9),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app10_parser = h_struct(
        app10_t,
        h_word_u16be(APP10),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app11_parser = h_struct(
        app11_t,
        h_word_u16be(APP11),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app12_parser = h_struct(
        app12_t,
        h_word_u16be(APP12),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app13_parser = h_struct(
        app13_t,
        h_word_u16be(APP13),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app14_parser = h_struct(
        app14_t,
        h_word_u16be(APP14),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *app15_parser = h_struct(
        app15_t,
        h_word_u16be(APP15),
        h_word_u16be(),
        h_bytes_until(h_any())
    );

    HParser *jpeg_parser = h_choice(
        soi_parser,
        app0_parser,
        app1_parser,
        dqt_parser,
        sof0_parser,
        dht_parser,
        sos_parser,
        eoi_parser,
        dri_parser,
        com_parser,
        app2_parser,
        app3_parser,
        app4_parser,
        app5_parser,
        app6_parser,
        app7_parser,
        app8_parser,
        app9_parser,
        app10_parser,
        app11_parser,
        app12_parser,
        app13_parser,
        app14_parser,
        app15_parser
    );

    HParseResult *result = h_parse(jpeg_parser, data, file_size);
    if (result->status == H_OK) {
        printf("Parsed JPEG file successfully\n");
    } else {
        printf("Error parsing JPEG file\n");
    }

    free(data);
    return 0;
}

File: results/0.75/JPEG/jpeg-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    bytes::complete::{tag, take, take_while_m_n},
    combinator::{map, opt},
    number::complete::{be_u16, be_u8},
    sequence::{tuple},
    IResult,
    error::ErrorKind,
};
use std::{env, fs, result};

#[derive(Debug, PartialEq)]
enum Marker {
    SOI,
    APP0,
    APP1,
    DQT,
    DHT,
    SOF,
    DRI,
    SOS,
    RST(u8),
    EOI,
}

fn marker(input: &[u8]) -> IResult<&[u8], Marker> {
    let (input, _) = tag(&[0xFF])(input)?;
    let (input, marker) = be_u8(input)?;
    match marker {
        0xD8 => Ok((input, Marker::SOI)),
        0xE0 => Ok((input, Marker::APP0)),
        0xE1 => Ok((input, Marker::APP1)),
        0xDB => Ok((input, Marker::DQT)),
        0xC4 => Ok((input, Marker::DHT)),
        0xC0 => Ok((input, Marker::SOF)),
        0xDD => Ok((input, Marker::DRI)),
        0xDA => Ok((input, Marker::SOS)),
        0xD0..=0xD7 => Ok((input, Marker::RST(marker - 0xD0))),
        0xD9 => Ok((input, Marker::EOI)),
        _ => Err(nom::Err::Error(nom::error::Error::new(input, ErrorKind::AlphaNumeric))),
    }
}

fn app0(input: &[u8]) -> IResult<&[u8], (&[u8], u8, u8, u8, u8, u8)> {
    let (input, _) = tag(&[0xFF, 0xE0])(input)?;
    let (input, length) = be_u16(input)?;
    let (input, identifier) = take(5u8)(input)?;
    let (input, version) = tuple((be_u8, be_u8))(input)?;
    let (input, units) = be_u8(input)?;
    let (input, _x_density) = be_u8(input)?;
    let (input, _y_density) = be_u8(input)?;
    let (input, x_thumbnail) = be_u8(input)?;
    let (input, y_thumbnail) = be_u8(input)?;
    Ok((
        &input[length as usize - 8..],
        (
            identifier,
            version.0,
            version.1,
            units,
            x_thumbnail,
            y_thumbnail,
        ),
    ))
}

fn dqt(input: &[u8]) -> IResult<&[u8], (u8, u8, Vec<u8>)> {
    let (input, _) = tag(&[0xFF, 0xDB])(input)?;
    let (input, length) = be_u16(input)?;
    let (input, table_number) = be_u8(input)?;
    let (input, precision) = be_u8(input)?;
    let (input, table_data) = take(length - 3)(input)?;
    Ok((
        &input[length as usize - 3..],
        (table_number, precision, table_data.to_vec()),
    ))
}

fn dht(input: &[u8]) -> IResult<&[u8], (u8, u8, Vec<u8>)> {
    let (input, _) = tag(&[0xFF, 0xC4])(input)?;
    let (input, length) = be_u16(input)?;
    let (input, table_class) = be_u8(input)?;
    let (input, table_destination_identifier) = be_u8(input)?;
    let (input, _number_of_codes) = be_u8(input)?;
    let (input, code_lengths) = take(length - 3)(input)?;
    Ok((
        &input[length as usize - 3..],
        (
            table_class,
            table_destination_identifier,
            code_lengths.to_vec(),
        ),
    ))
}

fn sof(input: &[u8]) -> IResult<&[u8], (u8, u16, u16, u8)> {
    let (input, _) = tag(&[0xFF, 0xC0])(input)?;
    let (input, length) = be_u16(input)?;
    let (input, precision) = be_u8(input)?;
    let (input, image_height) = be_u16(input)?;
    let (input, image_width) = be_u16(input)?;
    let (input, number_of_components) = be_u8(input)?;
    Ok((
        &input[length as usize - 8..],
        (precision, image_height, image_width, number_of_components),
    ))
}

fn dri(input: &[u8]) -> IResult<&[u8], u16> {
    let (input, _) = tag(&[0xFF, 0xDD])(input)?;
    let (input, length) = be_u16(input)?;
    let (input, restart_interval) = be_u16(input)?;
    Ok((&input[length as usize - 2..], restart_interval))
}

fn sos(input: &[u8]) -> IResult<&[u8], (u8, Vec<u8>)> {
    let (input, _) = tag(&[0xFF, 0xDA])(input)?;
    let (input, length) = be_u16(input)?;
    let (input, number_of_components) = be_u8(input)?;
    let (input, component_identifiers) = take(length - 3)(input)?;
    Ok((
        &input[length as usize - 3..],
        (number_of_components, component_identifiers.to_vec()),
    ))
}

fn rst(input: &[u8]) -> IResult<&[u8], u8> {
    let (input, _) = tag(&[0xFF, 0xD0])(input)?;
    let (input, rst_number) = be_u8(input)?;
    Ok((&input[1..], rst_number - 0xD0))
}

fn eoi(input: &[u8]) -> IResult<&[u8], ()> {
    let (input, _) = tag(&[0xFF, 0xD9])(input)?;
    Ok((input, ()))
}

fn main() -> result::Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        return Err("Invalid arguments".into());
    }
    let input_file = &args[1];
    let data = match fs::read(input_file) {
        Ok(data) => data,
        Err(err) => {
            eprintln!("Error reading file: {}", err);
            return Err(err.into());
        }
    };
    let mut input = &data[..];
    loop {
        match marker(input) {
            Ok((new_input, marker)) => {
                input = new_input;
                match marker {
                    Marker::SOI => println!("SOI"),
                    Marker::APP0 => {
                        let (new_input, app0) = match app0(input) {
                            Ok(result) => result,
                            Err(err) => {
                                eprintln!("Error parsing APP0: {:?}", err);
                                break;
                            }
                        };
                        input = new_input;
                        println!("APP0 {:?}", app0);
                    }
                    Marker::APP1 => {
                        let (new_input, _) = take_while_m_n::<_, &[u8], nom::error::Error<&[u8]>>(2, 65535, |c| c != 0xFF)(input)?;
                        input = new_input;
                    }
                    Marker::DQT => {
                        let (new_input, dqt) = match dqt(input) {
                            Ok(result) => result,
                            Err(err) => {
                                eprintln!("Error parsing DQT: {:?}", err);
                                break;
                            }
                        };
                        input = new_input;
                        println!("DQT {:?}", dqt);
                    }
                    Marker::DHT => {
                        let (new_input, dht) = match dht(input) {
                            Ok(result) => result,
                            Err(err) => {
                                eprintln!("Error parsing DHT: {:?}", err);
                                break;
                            }
                        };
                        input = new_input;
                        println!("DHT {:?}", dht);
                    }
                    Marker::SOF => {
                        let (new_input, sof) = match sof(input) {
                            Ok(result) => result,
                            Err(err) => {
                                eprintln!("Error parsing SOF: {:?}", err);
                                break;
                            }
                        };
                        input = new_input;
                        println!("SOF {:?}", sof);
                    }
                    Marker::DRI => {
                        let (new_input, dri) = match dri(input) {
                            Ok(result) => result,
                            Err(err) => {
                                eprintln!("Error parsing DRI: {:?}", err);
                                break;
                            }
                        };
                        input = new_input;
                        println!("DRI {:?}", dri);
                    }
                    Marker::SOS => {
                        let (new_input, sos) = match sos(input) {
                            Ok(result) => result,
                            Err(err) => {
                                eprintln!("Error parsing SOS: {:?}", err);
                                break;
                            }
                        };
                        input = new_input;
                        println!("SOS {:?}", sos);
                    }
                    Marker::RST(rst_number) => {
                        let (new_input, _) = rst(input)?;
                        input = new_input;
                        println!("RST {:?}", rst_number);
                    }
                    Marker::EOI => {
                        let (new_input, _) = eoi(input)?;
                        input = new_input;
                        println!("EOI");
                        break;
                    }
                }
            }
            Err(err) => {
                eprintln!("Error parsing marker: {:?}", err);
                break;
            }
        }
    }
    Ok(())
}